### DISCUSSION TITLE
questionnaire export format

### DISCUSSION THREAD

**Discussion by BenGardiner** [strictdoc-project/strictdoc#2001] [Ideas]
Hello amazing strictdoc people :)

We are planning to migrate https://github.com/nmfta-repo/nmfta-telematics_security_requirements away from excel to strictdoc (as mentioned previously in https://github.com/strictdoc-project/strictdoc/discussions/604) ; this will let us manage all those telematics security requirements without meticulously grooming the excel printable; _IF_ we can also get the questionnaire format used there generated by strictdoc.

The questionnaire format there is one sheet per the 4x solution components -- modelled in strictdoc as refined requirements in each of 4 document fragments (see screen shot here):
![image](https://github.com/user-attachments/assets/ffca933c-5e06-47fb-92db-90e7ce4715bc)

Each sheet is one row per applicable requirement, sorted by criticality (of the parent requirement in this case), with a response section for each requirement

```
Response
--
Yes | In-Part | No | N/A | Notes
```

I think excel (or other spreadsheet format) is most appropriate because it is easiest for suppliers to fill-out and for procurement to collect and compare responses as well.  I have some experience modifying excel features to strictdoc some time ago, I think I know how to make the unit tests and format converters. 

I don't really know how I will approach:
1. creating one sheet per fragment in a re-usable / user-configurable way
2. sorting by the criticality of the parent requirement: neither how to look up the 'inherited' value nor how to do this in a re-usable / user-configurable way

**Comment by stanislaw** [comment]
Hi @BenGardiner 😄 

It's great to see that you are using strictdoc seriously!

I think I understand what you are trying to do.

Here are a few snippets that could be relevant for your task:

1) You use `DocumentCachingIterator` to iterate over the nodes of a document:

```
            document_iterator = DocumentCachingIterator(document)
            d_01_document_iterators[document] = document_iterator

            for node in document_iterator.all_content(
                print_fragments=False,
                print_fragments_from_files=False,
            ):
                if isinstance(node, SDocNode):
```

This way you can connect all nodes to a plain list and then sort it using `the criticality of the parent requirement`.

2) To get a parent requirement of a requirement you can use:

```
                    traceability_index.graph_database.get_link_value(
                        link_type=GraphLinkType.UID_TO_REQUIREMENT_CONNECTIONS,
                        lhs_node=requirement_id_,
                    )
```

The `SDocNodeConnections` class holds the parent and child nodes for a given node.

3) Again, the `DocumentCachingIterator` is a good example of how the included documents are iterated inside an including document.

4) And finally, the Excel writer used by StrictDoc supports sheets, see `worksheet = workbook.add_worksheet(name=EXCEL_SHEET_NAME)` and then `worksheet.write`, so you should be able to create your sheets, one sheet per fragment.

I think from this puzzle it should be possible to build your solution but let me know if it is approachable given your knowledge of strictdoc.

Cheers!

**Comment by BenGardiner** [comment]
Thanks @stanislaw !

That was the perfect primer to get me going. I authored a custom export script for this because it didn't seem like a new export type that you would be seeking. But in case the extra coverage could be useful in tests I proposed it as a `tests/integration/scripting_examples` in a PR: https://github.com/strictdoc-project/strictdoc/pull/2034
